### ISSUE TITLE
False Sense of Safety When Using “this” Keyword

### ISSUE THREAD

**Issue by john-larson** [Microsoft/TypeScript#30073]
“`this`” keyword in Javascript is context dependent, which means its value depends on the calling function. And this is one of the culprits of most subtle run-time errors in Javascript.
On the other hand, Typescript gives a different meaning to it and uses “`this`” keyword to refer to class properties from inside class methods.

But the problem is, when “`this`” is used in callback functions or in functions given to foreach as argument, IDE does not raise any compile-time errors, giving us the false sense of security, but we get run-time errors because “`this`” is undefined.

There seem to be two work-arounds:

1.	Using arrow functions
2.	Using `.bind(this)` syntax

I believe that it should not be necessary to use these work-arounds and Typescript should be doing correct static type checking of “`this`” as long as we are not disabling it by using “`any`” etc. Because it is the main reason we are using Typescript.

I do not have much idea about the inner workings of Typescript, but I assume this could be accomplished by doing a “`let self = this;`” kind of assignment in the transpiling process and using this variable when “this” is used in code.

**Comment by j-oliveras** [comment]
If I understand correctly, probably the option `noImplicitThis` solve most of the problems with `this` that you describe:
```ts
function aaaaa() {
  this.z = 1;
}

[1, 2, 3, 4].map(function () { this.x = 0; })
```
[Playground](https://www.typescriptlang.org/play/#src=function%20aaaaa()%20%7B%0D%0A%20%20this.z%20%3D%201%3B%0D%0A%7D%0D%0A%0D%0A%5B1%2C%202%2C%203%2C%204%5D.map(function%20()%20%7B%20this.x%20%3D%200%3B%20%7D)%0D%0A) (enable the option under Option button).

With that option enabled, flags both `this` usages with `'this' implicitly has type 'any' because it does not have a type annotation.`.

See [compiler options](https://www.typescriptlang.org/docs/handbook/compiler-options.html).

You can define the type of this inside a function with [that syntax](https://www.typescriptlang.org/docs/handbook/functions.html#this-parameters):
```ts
function aaa(this: { x: number }) {
  this.x = 1;
}
```
[Playground](https://www.typescriptlang.org/play/#src=function%20aaa(this%3A%20%7B%20x%3A%20number%20%7D)%20%7B%0D%0A%20%20this.x%20%3D%201%3B%0D%0A%7D%0D%0A).

**Comment by john-larson** [comment]
Enabling `noImplicitThis` will only force you to define `this `as a function argument with its type but this does not solve the problem. Compiler will not complain but `this `will still be `undefined `for 'use strict' and `window `otherwise.

Here is an example:

[Playground](https://www.typescriptlang.org/play/index.html#src=class%20SomeClass%7B%0D%0A%0D%0A%20%20log(arg)%20%7B%0D%0A%20%20%20%20console.log(arg)%3B%0D%0A%20%20%7D%0D%0A%0D%0A%20%20fe()%20%7B%0D%0A%20%20%20%20var%20list%20%3D%20%5B0%2C%201%2C%202%2C%203%5D%3B%0D%0A%0D%0A%20%20%20%20list.forEach(function%20(this%3A%20SomeClass%2C%20listItem)%20%7B%0D%0A%20%20%20%20%20%20console.log(this)%3B%0D%0A%20%20%20%20%20%20this.log(listItem)%3B%0D%0A%20%20%7D)%0D%0A%20%20%7D%0D%0A%0D%0A%20%20fe2()%20%7B%0D%0A%20%20%20%20var%20list%20%3D%20%5B0%2C%201%2C%202%2C%203%5D%3B%0D%0A%0D%0A%20%20%20%20list.forEach((listItem)%20%3D%3E%20%7B%0D%0A%20%20%20%20%20%20console.log(this)%3B%0D%0A%20%20%20%20%20%20this.log(listItem)%3B%0D%0A%20%20%20%20%7D)%0D%0A%20%20%7D%0D%0A%7D%0D%0A%0D%0Avar%20obj%20%3D%20new%20SomeClass()%3B%0D%0Aobj.fe()%3B)

Calling `obj.fe()` will not work whereas calling `fe2()` will work.

**Comment by DanielRosenwasser** [comment]
Issue #7968 covers strict `this` checks (called `--strictThis` when originally implemented in #6018).

**Comment by RyanCavanaugh** [comment]
See #513 and start following links from there 😉

**Comment by john-larson** [comment]
@RyanCavanaugh I certainly think this issue deserves to remain open as none of the items listed in #513 is related to what I am proposing here. I am proposing that `this` inside a class method should correctly be referencing the class instance as is expected. (I say as expected because it is what the transpiler assumes without options like `--strictThis` and it is also what developers assume since you need to use `this` in order to reference class properties from inside the methods). Feel free to reference this issue from other summary issue lists but I think this issue should be reopened.

**Comment by RyanCavanaugh** [comment]
> I am proposing that `this` inside a class method should always be correctly referencing the class instance as is expected

Classes are a JavaScript feature. TypeScript does not alter the runtime behavior of JavaScript, including behavior around `this`.
